// ol/format/gml2.js
import _ol_ from '../index';
import _ol_extent_ from '../extent';
import _ol_format_GMLBase_ from '../format/gmlbase';
import _ol_format_XSD_ from '../format/xsd';
import _ol_proj_ from '../proj';
import _ol_xml_ from '../xml';

/**
 * @classdesc
 * Feature format for reading and writing data in the GML format,
 * version 2.1.2.
 *
 * @constructor
 * @param {olx.format.GMLOptions=} opt_options Optional configuration object.
 * @extends {ol.format.GMLBase}
 * @api
 */
var _ol_format_GML2_ = function(opt_options) {
  var options = /** @type {olx.format.GMLOptions} */
      (opt_options ? opt_options : {});

  _ol_format_GMLBase_.call(this, options);

  this.FEATURE_COLLECTION_PARSERS[_ol_format_GMLBase_.GMLNS][
      'featureMember'] =
      _ol_xml_.makeArrayPusher(_ol_format_GMLBase_.prototype.readFeaturesInternal);

  /**
   * @inheritDoc
   */
  this.schemaLocation = options.schemaLocation ?
      options.schemaLocation : _ol_format_GML2_.schemaLocation_;

};

_ol_.inherits(_ol_format_GML2_, _ol_format_GMLBase_);


/**
 * @const
 * @type {string}
 * @private
 */
_ol_format_GML2_.schemaLocation_ = _ol_format_GMLBase_.GMLNS +
    ' http://schemas.opengis.net/gml/2.1.2/feature.xsd';


/**
 * @param {Node} node Node.
 * @param {Array.<*>} objectStack Object stack.
 * @private
 * @return {Array.<number>|undefined} Flat coordinates.
 */
_ol_format_GML2_.prototype.readFlatCoordinates_ = function(node, objectStack) {
  var s = _ol_xml_.getAllTextContent(node, false).replace(/^\s*|\s*$/g, '');
  var context = /** @type {ol.XmlNodeStackItem} */ (objectStack[0]);
  var containerSrs = context['srsName'];
  var containerDimension = node.parentNode.getAttribute('srsDimension');
  var axisOrientation = 'enu';
  if (containerSrs) {
    var proj = _ol_proj_.get(containerSrs);
    if (proj) {
      axisOrientation = proj.getAxisOrientation();
    }
  }
  var coords = s.split(/[\s,]+/);
  // The "dimension" attribute is from the GML 3.0.1 spec.
  var dim = 2;
  if (node.getAttribute('srsDimension')) {
    dim = _ol_format_XSD_.readNonNegativeIntegerString(
        node.getAttribute('srsDimension'));
  } else if (node.getAttribute('dimension')) {
    dim = _ol_format_XSD_.readNonNegativeIntegerString(
        node.getAttribute('dimension'));
  } else if (containerDimension) {
    dim = _ol_format_XSD_.readNonNegativeIntegerString(containerDimension);
  }
  var x, y, z;
  var flatCoordinates = [];
  for (var i = 0, ii = coords.length; i < ii; i += dim) {
    x = parseFloat(coords[i]);
    y = parseFloat(coords[i + 1]);
    z = (dim === 3) ? parseFloat(coords[i + 2]) : 0;
    if (axisOrientation.substr(0, 2) === 'en') {
      flatCoordinates.push(x, y, z);
    } else {
      flatCoordinates.push(y, x, z);
    }
  }
  return flatCoordinates;
};


/**
 * @param {Node} node Node.
 * @param {Array.<*>} objectStack Object stack.
 * @private
 * @return {ol.Extent|undefined} Envelope.
 */
_ol_format_GML2_.prototype.readBox_ = function(node, objectStack) {
  /** @type {Array.<number>} */
  var flatCoordinates = _ol_xml_.pushParseAndPop([null],
      this.BOX_PARSERS_, node, objectStack, this);
  return _ol_extent_.createOrUpdate(flatCoordinates[1][0],
      flatCoordinates[1][1], flatCoordinates[1][3],
      flatCoordinates[1][4]);
};


/**
 * @param {Node} node Node.
 * @param {Array.<*>} objectStack Object stack.
 * @private
 */
_ol_format_GML2_.prototype.innerBoundaryIsParser_ = function(node, objectStack) {
  /** @type {Array.<number>|undefined} */
  var flatLinearRing = _ol_xml_.pushParseAndPop(undefined,
      this.RING_PARSERS, node, objectStack, this);
  if (flatLinearRing) {
    var flatLinearRings = /** @type {Array.<Array.<number>>} */
        (objectStack[objectStack.length - 1]);
    flatLinearRings.push(flatLinearRing);
  }
};


/**
 * @param {Node} node Node.
 * @param {Array.<*>} objectStack Object stack.
 * @private
 */
_ol_format_GML2_.prototype.outerBoundaryIsParser_ = function(node, objectStack) {
  /** @type {Array.<number>|undefined} */
  var flatLinearRing = _ol_xml_.pushParseAndPop(undefined,
      this.RING_PARSERS, node, objectStack, this);
  if (flatLinearRing) {
    var flatLinearRings = /** @type {Array.<Array.<number>>} */
        (objectStack[objectStack.length - 1]);
    flatLinearRings[0] = flatLinearRing;
  }
};


/**
 * @const
 * @type {Object.<string, Object.<string, ol.XmlParser>>}
 * @private
 */
_ol_format_GML2_.prototype.GEOMETRY_FLAT_COORDINATES_PARSERS_ = {
  'http://www.opengis.net/gml': {
    'coordinates': _ol_xml_.makeReplacer(
        _ol_format_GML2_.prototype.readFlatCoordinates_)
  }
};


/**
 * @const
 * @type {Object.<string, Object.<string, ol.XmlParser>>}
 * @private
 */
_ol_format_GML2_.prototype.FLAT_LINEAR_RINGS_PARSERS_ = {
  'http://www.opengis.net/gml': {
    'innerBoundaryIs': _ol_format_GML2_.prototype.innerBoundaryIsParser_,
    'outerBoundaryIs': _ol_format_GML2_.prototype.outerBoundaryIsParser_
  }
};


/**
 * @const
 * @type {Object.<string, Object.<string, ol.XmlParser>>}
 * @private
 */
_ol_format_GML2_.prototype.BOX_PARSERS_ = {
  'http://www.opengis.net/gml': {
    'coordinates': _ol_xml_.makeArrayPusher(
        _ol_format_GML2_.prototype.readFlatCoordinates_)
  }
};


/**
 * @const
 * @type {Object.<string, Object.<string, ol.XmlParser>>}
 * @private
 */
_ol_format_GML2_.prototype.GEOMETRY_PARSERS_ = {
  'http://www.opengis.net/gml': {
    'Point': _ol_xml_.makeReplacer(_ol_format_GMLBase_.prototype.readPoint),
    'MultiPoint': _ol_xml_.makeReplacer(
        _ol_format_GMLBase_.prototype.readMultiPoint),
    'LineString': _ol_xml_.makeReplacer(
        _ol_format_GMLBase_.prototype.readLineString),
    'MultiLineString': _ol_xml_.makeReplacer(
        _ol_format_GMLBase_.prototype.readMultiLineString),
    'LinearRing': _ol_xml_.makeReplacer(
        _ol_format_GMLBase_.prototype.readLinearRing),
    'Polygon': _ol_xml_.makeReplacer(_ol_format_GMLBase_.prototype.readPolygon),
    'MultiPolygon': _ol_xml_.makeReplacer(
        _ol_format_GMLBase_.prototype.readMultiPolygon),
    'Box': _ol_xml_.makeReplacer(_ol_format_GML2_.prototype.readBox_)
  }
};
export default _ol_format_GML2_;
